### PR TITLE
cincinnati/plugins/edge_add_remove: graceful and consistent handling of all inconsistent label constellations

### PR DESCRIPTION
**Pull request by steveej**
Introduces consistency in gracefully handle all expected inconsistent label constellations for edge addition and removal.

Please see the commit messages for more details.

### CODE REVIEW

File: cincinnati/src/lib.rs [hunk 1]
```diff
@@ -145,6 +146,17 @@ impl<'a> Iterator for PreviousReleases<'a> {
 #[derive(Debug, Clone)]
 pub struct Empty;
 
+/// Errors that can be returned by the methods in this library
+pub mod errors {
+    /// Edge already exists
+    #[derive(Debug, Fail, Eq, PartialEq)]
+    #[fail(display = "edge from {:?} to {:?} already exists", from, to)]
```
Comment (lucab) [inline cincinnati/src/lib.rs:152]: I'm not sure this is very helpful. Node indexes depend on the specific graph instance/ordering. Should we maybe track source and destination versions instead?

File: cincinnati/src/plugins/internal/edge_add_remove.rs [hunk 2]
```diff
@@ -128,18 +132,30 @@ impl EdgeAddRemovePlugin {
...
+
         graph
             .find_by_metadata_key(&format!("{}.{}", self.key_prefix, "previous.add"))
             .into_iter()
-            .try_for_each(|(to, to_version, from_csv)| -> Fallible<()> {
-                for from_version in from_csv.split(',').map(str::trim) {
-                    if let Some(from) = graph.find_by_version(&from_version) {
-                        info!("[{}]: adding previous {}", &from_version, &to_version);
-                        graph.add_edge(&from, &to)?
+            .try_for_each(|(to, to_string, from_csv)| -> Fallible<()> {
```
Comment (lucab) [inline cincinnati/src/plugins/internal/edge_add_remove.rs:166]: It looks like this `try_for_each` was drifting rightward and you had to introduce a macro. Did you maybe consider splitting this whole closure to its own function?

File: commons/src/testing.rs [hunk 3]
```diff
@@ -3,9 +3,15 @@
 use failure::Fallible;
 use tokio::runtime::current_thread::Runtime;
 
+/// Initialize logging.
+pub fn init_logger() -> Fallible<()> {
+    env_logger::try_init_from_env(env_logger::Env::default())?;
```
Comment (lucab) [inline commons/src/testing.rs:8]: The original location where you moved this from didn't have the error unwrapping. I think the `?` in here will make this failing as soon as there two tests using it, as the logger for subsequent tests is already initialized by the first running test.

Comment (steveej) [reply commons/src/testing.rs:8]: That's right, I didn't want to swallow the error here. The caller can if it wants to. Do you prefer an opinionated abstraction here?

File: cincinnati/src/plugins/internal/edge_add_remove.rs (hunk 2 above)
Comment (steveej) [reply cincinnati/src/plugins/internal/edge_add_remove.rs:166]: I tried to use the macro to reduce the duplication between the closures for the previous and next. A function wouldn't help with that use-case, and I wasn't sure how much effort I want to put into making one macro for both cases.

File: cincinnati/src/lib.rs (hunk 1 above)
Comment (steveej) [reply cincinnati/src/lib.rs:152]: I changed it to be a String here so it can be fed a version string at instantiation.

**Comment by steveej** [comment]
@lucab I increased the scope of this PR a bit as I kept hitting more error cases in the tests and wanted them handled properly. Please take a look at the new tests and the commit messages for a complete understanding.

File: commons/src/testing.rs (hunk 3 above)
Comment (lucab) [reply commons/src/testing.rs:8]: If that was on purpose and we plan to stick with swallowing error at caller site, I'm fine with keeping this.

File: cincinnati/src/plugins/internal/edge_add_remove.rs (hunk 2 above)
Comment (lucab) [reply cincinnati/src/plugins/internal/edge_add_remove.rs:166]: I guess we can leave this as-is for now then (I think that my minor itches here are the two macro-in-function).

If/when we need to have the same permissive add/remove logic in some other place, we can reconsider how to better share this logic.

File: cincinnati/src/plugins/internal/edge_add_remove.rs [hunk 4]
```diff
@@ -46,9 +46,17 @@ impl PluginSettings for EdgeAddRemovePlugin {
...
+///     1. previous
+///     2. next
 /// 2. *.remove
+///     1. previous
+///     2. next
 ///
 /// This ordering has implications on the result of semantical contradictions, so that the `*.remove` labels take precedence over `*.add`.
+///
+/// # Strictness
+/// The plugin aims to gracefully handle any inconstistencies to make the operation as robust as possible.
```
Comment (lucab) [inline cincinnati/src/plugins/internal/edge_add_remove.rs]: Typo: `inconsistencies`

File: cincinnati/src/lib.rs [hunk 5]
```diff
@@ -145,6 +146,25 @@ impl<'a> Iterator for PreviousReleases<'a> {
...
+    pub struct EdgeAlreadyExists {
+        pub(crate) from: String,
+        pub(crate) to: String,
+    }
+
+    /// Edge already exists
```
Comment (lucab) [inline cincinnati/src/lib.rs]: I think this is a copy-paste mismatch.

**Comment by lucab** [comment]
@steveeJ thanks, I think it makes sense to have them all here as they fall under the same logical umbrella. I made another pass, and I only spotted minor typos.

**Comment by steveej** [comment]
@lucab Thanks for the review. I addressed your comments and also removed two commits which added unused code and the lack of a good test use-case. Please take another look and stamp if you think it's good.

**Review by lucab — APPROVED** [review]

**Pull request merged**